### DISCUSSION TITLE
Delegate in function parameters

### DISCUSSION THREAD

**Discussion by Aragas** [elringus/bootsharp#78] [Q&A]
Hey!
I'm wondering what would be the best way to expose C# functions to JavaScript that need delegates passed as parameters?
```csharp
[JSInvokable]
string Test(string param, Func<string, string> request, Action<string> action) => string.Empty;
```

I guess passing `Action`/`Func` is not an option, should an `event` approach be used then?
```csharp
[JSEvent]
public static partial string Request(string payload);

[JSEvent]
public static partial void Action(string payload);

[JSInvokable]
string Test(string param) => string.Empty;
```
The scope will be lost with this approach, is it possible to create a nested class?
```csharp
public class Scope
{
    [JSEvent]
    public static partial string Request(string payload);

    [JSEvent]
    public static partial void Action(string payload);

    [JSInvokable]
    string Test(string param) => string.Empty;
}

[JSInvokable]
DotNetObjectReference<Scope> CreateScope() => new Scope();
```

**Comment by elringus** [comment]
That's in the domain of .NET JS interop. Afaik, delegates can't be passed directly, but instances wrapped in `DotNetObjectReference` can. https://docs.microsoft.com/en-us/aspnet/core/blazor/javascript-interoperability/call-dotnet-from-javascript?view=aspnetcore-6.0#pass-a-dotnetobjectreference-to-an-individual-javascript-function
